File: exams/1617/translate/argparse/solution/argparse/parser.cpp
#include "parser.h"


Parser& Parser::add_parameter(std::shared_ptr<Parameter> parameter)
{
	m_parameters.push_back(parameter);

	return *this;
}

bool Parser::process(std::list<std::string>& arguments) const
{
	for (auto& parameter : m_parameters)
	{
		if (parameter->parse(arguments))
		{
			return true;
		}
	}

	return false;
}

void Parser::parse(std::list<std::string>& arguments) const
{
	std::list<std::string> positional;

	while (!arguments.empty())
	{
		if (!process(arguments))
		{
			positional.push_back(arguments.front());
			arguments.pop_front();
		}
	}

	arguments = positional;
}

File: exams/1617/translate/argparse/student/java/Parser.java
import java.util.ArrayList;


class Parser
{
    private ArrayList<Parameter> parameters;

    public Parser()
    {
        this.parameters = new ArrayList<Parameter>();
    }
    
    private boolean process(ArrayList<String> args)
    {
        for ( Parameter parameter : parameters )
        {
            if ( parameter.parse(args) )
            {
                return true;
            }
        }

        return false;
    }

    public Parser addParameter(Parameter parameter)
    {
        parameters.add(parameter);

        return this;
    }

    public void parse(ArrayList<String> args)
    {
        ArrayList<String> positional = new ArrayList<>();

        while ( !args.isEmpty() )
        {
            if ( !process(args) )
            {
                positional.add(args.get(0));
                args.remove(0);
            }
        }

        args.addAll(positional);
    }    
}
